ambry-router/src/main/java/com.github.ambry.router/PutOperation.java

```diff
@@ -353,8 +353,12 @@ void fillChunks() {
       if (chunkFillingCompletedSuccessfully) {
         PutChunk lastChunk = getBuildingChunk();
         if (lastChunk != null) {
-          lastChunk.onFillComplete(true);
-          updateChunkFillerWaitTimeMetrics();
+          if (chunkCounter != 0 && lastChunk.buf.position() == 0) {
```

Comment:
As we saw in person, `lastChunk` refers to the `PutChunk` and not the buffer from the `ReadableStreamChannel` (which is named as `channelReadBuffer` in this method). For the `PutChunk` buffer, this check is guaranteed to be correct as this is done only if the chunk is in `Building` state and the method that takes the chunk to that state (`prepareForBuilding()`) sets the position to 0.

For the second part, yes - `testDelayedChannelPutSuccess()` in `PutManagerTest` tests zero-sized last as well as intermediate buffers (probabilistically for the latter).